### PR TITLE
clp-package: Add support for printing real-time compression statistics.

### PR DESCRIPTION
**Pull request by wraymo**
# References

# Description

For large jobs, users often experience extended waiting periods until completion. During that time, they have no idea about the compression status. This PR addresses this issue by printing out real-time statistics (compression ratio and speed) during compression. Specifically, this PR introduces the code changes listed below:
+ `CompressionTaskFailureResult` and `CompressionTaskSuccessResult` are removed.
+ The compression scheduler does not update the completion status of compression tasks.
+ Instead, the completion status of compression tasks are updated within each task and the fields (i.e. `compressed_size`, `uncompressed_size` and `num_tasks_completed`) of compression jobs are updated at the same time.
+ For the user compression script, `completion_query` is removed and compression statistics are now printed promptly upon their update.

# Validation performed

+ Built and started the package
+ Compressed the [hive-24hrs](https://zenodo.org/records/7094921#.Y5JbH33MKHs) dataset. It printed out statistics like this
```
2024-05-10 03:22:52,628 [INFO] [/opt/clp/lib/python3/site-packages/clp_package_utils/scripts/native/compress.py] Compression job 6 submitted.
2024-05-10 03:23:02,159 [INFO] [/opt/clp/lib/python3/site-packages/clp_package_utils/scripts/native/compress.py] Compressed 159.87MB into 4.14MB (38.58). Speed: 48.95MB/s.
2024-05-10 03:23:04,165 [INFO] [/opt/clp/lib/python3/site-packages/clp_package_utils/scripts/native/compress.py] Compressed 928.14MB into 23.46MB (39.56). Speed: 176.05MB/s.
2024-05-10 03:23:04,667 [INFO] [/opt/clp/lib/python3/site-packages/clp_package_utils/scripts/native/compress.py] Compressed 1.41GB into 36.32MB (39.73). Speed: 249.91MB/s.
2024-05-10 03:23:05,670 [INFO] [/opt/clp/lib/python3/site-packages/clp_package_utils/scripts/native/compress.py] Compressed 1.66GB into 40.11MB (42.36). Speed: 250.69MB/s.
2024-05-10 03:23:06,673 [INFO] [/opt/clp/lib/python3/site-packages/clp_package_utils/scripts/native/compress.py] Compression finished. Compressed 1.99GB into 44.37MB (45.89). Speed: 261.66MB/s.
```

### CODE REVIEW

**Comment by kirkrodrigues** [comment]
Can you summarize the approach taken? I.e., at a high level, what is the code change.

**Comment by wraymo** [comment]
> Can you summarize the approach taken? I.e., at a high level, what is the code change.

Updated

File: components/job-orchestration/job_orchestration/executor/compress/fs_compression_task.py [hunk 1]
```diff
@@ -169,45 +202,37 @@ def run_clp(
...
+    # Compute the total amount of data compressed
+    last_archive_id = ""
+    total_uncompressed_size = 0
+    total_compressed_size = 0
+    while True:
+        line = proc.stdout.readline()
+        if not line:
+            break
+        stats = json.loads(line.decode("ascii"))
+        if stats["id"] != last_archive_id:
```
Comment (kirkrodrigues) [inline components/job-orchestration/job_orchestration/executor/compress/fs_compression_task.py:214]: Hm, since an archive may print its stats multiple times (after every segment is created in `clp`'s case), then we would still need the previous logic that keeps `last_archive_stats`, right?

File: components/job-orchestration/job_orchestration/executor/compress/fs_compression_task.py [hunk 2]
```diff
@@ -256,37 +283,53 @@ def compress(
...
-            task_id=task_id,
-            status=CompressionTaskStatus.SUCCEEDED,
-            start_time=start_time,
-            duration=duration,
-            total_uncompressed_size=worker_output["total_uncompressed_size"],
-            total_compressed_size=worker_output["total_compressed_size"],
-        ).dict()
-    else:
-        return CompressionTaskFailureResult(
+    with closing(sql_adapter.create_connection(True)) as db_conn, closing(
```
Comment (kirkrodrigues) [inline components/job-orchestration/job_orchestration/executor/compress/fs_compression_task.py:288]: I still think we should try to keep our MySQL connections short, even if frequent; so I would prefer we open a connection only just before we perform a write (currently, this opens a connection before we start compression, which itself could take a long time depending on the dataset and config).

MySQL's [blog](https://dev.mysql.com/blog-archive/mysql-connection-handling-and-scaling/) says that it's capable of handling a lot of short connections, but its default concurrent connection limit is only [151](https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_max_connections). We should probably benchmark this for ourselves at some point (a long time ago, @haiqi96 had done some scalability benchmarking that showed MySQL struggled with 20 concurrent connections performing inserts), but for now, I think following their advice is the safer option.

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]

File: components/job-orchestration/job_orchestration/executor/compress/fs_compression_task.py (hunk 2 above)
Comment (wraymo) [reply components/job-orchestration/job_orchestration/executor/compress/fs_compression_task.py:288]: Yeah, my original plan was to use short connection. But I notice in compression scheduler, we maintain a long connection.

File: components/job-orchestration/job_orchestration/executor/compress/fs_compression_task.py (hunk 1 above)
Comment (wraymo) [reply components/job-orchestration/job_orchestration/executor/compress/fs_compression_task.py:214]: The previous logic is to check last_archive_stats['id'], if it's different from current id, we update everything. That's the same as current logic?

Comment (kirkrodrigues) [reply components/job-orchestration/job_orchestration/executor/compress/fs_compression_task.py:214]: Not exactly. In the previous logic if `last_archive_stats['id']` was different from the current ID, we would add `uncompressed_size` and `size` from `last_archive_stats`, but in the current code, we're adding the values from `stats`.

To see why this is a problem, imagine `clp` creates two archives with two segments each, meaning it will print the archive stats 4 times, something like this:

1. archive-1-seg-1: `uncompressed_size = 10, size = 1`
2. archive-1-seg-2: `uncompressed_size = 20, size = 2`
3. archive-2-seg-1: `uncompressed_size = 5, size = 1`
4. archive-2-seg-2: `uncompressed_size = 10, size 2`

In the current code, when we see the printout of (1), we will do `total_uncompressed_size += 10, size += 1`. When we see the printout of (3), we will do `total_uncompressed_size += 5, size += 1`. This will give us `total_uncompressed_size = 15, size = 2`. But it should be `total_uncompressed_size = 30, size = 4`.

File: components/job-orchestration/job_orchestration/executor/compress/fs_compression_task.py (hunk 2 above)
Comment (kirkrodrigues) [reply components/job-orchestration/job_orchestration/executor/compress/fs_compression_task.py:288]: In the scheduler's case, it's only maintaining one connection that it's using for polling (among other things), right? In theory we could make it use shorter connections, but there I'm not sure it will make much difference (we should still measure at some point though).

Comment (wraymo) [reply components/job-orchestration/job_orchestration/executor/compress/fs_compression_task.py:288]: Got it. For the compression task, we read the output from the process and update the database. Do you think we should open a connection each time when we get a new line?

File: components/job-orchestration/job_orchestration/executor/compress/fs_compression_task.py (hunk 1 above)
Comment (wraymo) [reply components/job-orchestration/job_orchestration/executor/compress/fs_compression_task.py:214]: Oh I see your point. `uncompressed_size` is cumulative. I thought we can accept the first one, and abandon the rest with the same id.

File: components/job-orchestration/job_orchestration/executor/compress/fs_compression_task.py (hunk 2 above)
Comment (kirkrodrigues) [reply components/job-orchestration/job_orchestration/executor/compress/fs_compression_task.py:288]: I think we could try opening a connection each time we need to update the archive's stats + tags (which would only be every time we finish an archive) and then once at the end of the task.

File: components/job-orchestration/job_orchestration/executor/compress/fs_compression_task.py [hunk 3]
```diff
@@ -268,25 +320,38 @@ def compress(
...
-            start_time=start_time,
+            status=compression_task_status,
             duration=duration,
-            error_message=worker_output["error_message"],
-        ).dict()
+        )
+
+        if CompressionTaskStatus.SUCCEEDED == compression_task_status:
+            increment_compression_job_metadata(db_cursor, job_id, dict(num_tasks_completed=1))
+            db_conn.commit()
```
Comment (kirkrodrigues) [inline components/job-orchestration/job_orchestration/executor/compress/fs_compression_task.py:353]: Instead of separating the database ops, could we do all the database ops first and then prepare the CompressionTaskResult afterward, outside the `with` block? Something like:

```py
with ...:
  update_compression_task_metadata(...)

  if CompressionTaskStatus.SUCCEEDED == compression_task_status:
    increment_compression_job_metadata(...)

  db_conn.commit()

compression_task_result = ...
if CompressionTaskStatus.SUCCEEDED != compression_task_status:
  compression_task_result.error_message = ...
```

That would require repeating the `CompressionTaskStatus.SUCCEEDED` check, but I think it's a little clearer in terms of grouping operations.

File: components/clp-package-utils/clp_package_utils/scripts/native/compress.py [hunk 4]
```diff
@@ -35,23 +34,32 @@
...
 
+def print_compression_job_status(job_row, current_time):
+    job_uncompressed_size = job_row["uncompressed_size"]
+    job_compressed_size = job_row["compressed_size"]
+    job_start_time = job_row["start_time"]
+    compression_ratio = float(job_uncompressed_size) / job_compressed_size
+    compression_speed = job_uncompressed_size / (current_time - job_start_time).total_seconds()
+    logger.info(
+        f"Compressed {pretty_size(job_uncompressed_size)} into "
+        f"{pretty_size(job_compressed_size)} ({compression_ratio:.2f}). "
```
Comment (kirkrodrigues) [inline components/clp-package-utils/clp_package_utils/scripts/native/compress.py:45]: What do you think about adding `x` after the compression ratio?

File: components/job-orchestration/job_orchestration/executor/compress/fs_compression_task.py [hunk 5]
```diff
@@ -169,45 +212,50 @@ def run_clp(
...
-                        tag_ids,
-                    )
-            last_archive_stats = stats
-        if last_archive_stats is not None:
-            # Add the last archive's last reported size
+    # Compute the total amount of data compressed
+    last_archive_id = ""
+    total_uncompressed_size = 0
+    total_compressed_size = 0
+    last_archive_stats = None
```
Comment (kirkrodrigues) [inline components/job-orchestration/job_orchestration/executor/compress/fs_compression_task.py:219]: Very minor, but if you move this back above `total_uncompressed_size = 0`, the diff would be smaller.

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]

**Review by kirkrodrigues — APPROVED** [review]
